simulation/chart.py

```diff
@@ -1,232 +1,438 @@
-        size (float, optional): Size of the marker symbol. Defaults to None.
-        color (Any, optional): Color to fill the marker symbol. Defaults to None.
-        symbol (Any, optional): Shape of the marker symbol. Defaults to None.
-        border (Any, optional): Pen to draw border around the marker symbol. Defaults to None.
-    """
-    config = {}
 
-    if size is not None:
-        config["symbolSize"] = size
+class PenConfig(TypedDict, total=False):
+    """Configuration for pen to draw lines"""
 
-    if color is not None:
-        config["symbolBrush"] = color
+    width: float
+    """Width of the line"""
+    style: PenStyle
+    """Line style"""
+    color: Color
+    """Line color"""
 
-    if symbol is not None:
-        config["symbol"] = symbol
 
-    if border is not None:
-        config["symbolPen"] = border
+# Method to resample the data before plotting ot avoid plotting multiple line segments per pixel.
+DownsampleMethod = Literal[
+    "subsample",  # Downsample by taking the first of N samples. This method is fastest and least accurate.
+    "mean",  # Downsample by taking the mean of N samples.
+    "peak",  # Downsample by drawing a saw wave that follows the min and max of the original data. This method produces the best visual representation of the data but is slower.
+]
 
-    return chart.plot(name=name, pen=None, **config)
+DEFAULT_WIN_SIZE = (1000, 1000)
+DEFAULT_CODEC = "libx264"
+DEFAULT_PIXEL_FORMAT = "yuv420p"
 
 
-def create_video_process(
-    file_name: str,
-    size: Tuple[int, int],
-    codec: str = "libx264",
-    pixel_format: str = "yuv420p",
-) -> Popen:
-    """Creates a ffmpeg video process which accepts input from stdin.(NA in windows)
+class ChartWidget:
+    """A chart container for single group of plots and a legend.
 
     Args:
-        file_name (str): name of the file to save encoded video ouput
-        size (Tuple[int,int]): Size (width,height) of the video frame
-        codec (str, optional): Codec name to be used for encoding. Defaults to "libx264".
-        pixel_format (str, optional): Ouput pixel format. Defaults to "yuv420p".
-
-    Returns:
-        Process: Asyncronious sub process listening stdin for encoding to the file.
+        chart (PlotItem): base container plotitem
     """
-    process = (
-        ffmpeg.input(
-            "pipe:",
-            format="rawvideo",
-            pix_fmt="rgb24",
-            s=f"{size[0]}x{size[1]}",
+
+    def __init__(self, chart: pg.PlotItem) -> None:
+        self.__chart = chart
+        self.__plots: Dict[str, pg.PlotDataItem] = {}
+
+    @staticmethod
+    def _update_downsapling(
+        plot: pg.PlotDataItem,
+        auto: Optional[bool] = None,
+        ds: Optional[int] = None,
+        method: Optional[DownsampleMethod] = None,
+    ):
+        """Update plot downsampling parameters"""
+
+        if auto is None and ds is None and method is None:
+            return
+
+        config: Dict[str, Any] = {}
```

Comment:
This whole block with `None` checking doesn't make sense to me. Consider passing in the config dictionary instead and using `setdefault` to set values for keys that don't exist.

Reply:
Fully removed as same code is working on the pyqtgraph method `setDownsampling`.
https://github.com/semiotic-ai/autoagora-agents/commit/543ffd999da9d9e71b97c69fd84b59c0dec6a395